components/script/dom/resizeobserver.rs

```diff
@@ -0,0 +1,277 @@
+        &self,
+        depth: &ResizeObservationDepth,
+        has_active: &mut bool,
+    ) {
+        for (observation, target) in self.observation_targets.borrow_mut().iter_mut() {
+            observation.state = Default::default();
+            if observation.is_active(target) {
+                let target_depth = calculate_depth_for_node(target);
+                if target_depth > *depth {
+                    observation.state = ObservationState::Active;
+                    *has_active = true;
+                } else {
+                    observation.state = ObservationState::Skipped;
+                }
+            }
+        }
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#broadcast-active-resize-observations>
+    pub fn broadcast_active_resize_observations(
+        &self,
+        shallowest_target_depth: &mut ResizeObservationDepth,
+    ) {
+        let mut entries: Vec<DomRoot<ResizeObserverEntry>> = Default::default();
+        for (observation, target) in self.observation_targets.borrow_mut().iter_mut() {
+            if observation.state != ObservationState::Active {
+                continue;
+            }
+
+            // #create-and-populate-a-resizeobserverentry
+
+            // Note: only calculating content box size.
+            let box_size = calculate_box_size(target, &observation.observed_box);
+            let width = box_size.width().to_f64_px();
+            let height = box_size.height().to_f64_px();
+            let size_impl = ResizeObserverSizeImpl::new(width, height);
+            let window = window_from_node(&**target);
+            let observer_size = ResizeObserverSize::new(&*window, size_impl);
+
+            // Note: content rect is built from content box size.
+            let content_rect = DOMRectReadOnly::new(
+                &*window.upcast(),
+                None,
+                box_size.origin.x.to_f64_px(),
+                box_size.origin.y.to_f64_px(),
+                width,
+                height,
+            );
+            let entry = ResizeObserverEntry::new(
+                &*window,
+                target,
+                &*content_rect,
+                &[],
+                &[&*observer_size],
+                &[],
+            );
+            entries.push(entry);
+
+            // Note: this is safe because an observation is
+            // initialized with one reported size(zero).
+            // The spec plans to store multiple reported sizes,
+            // but for now there can be only one.
+            observation.last_reported_sizes[0] = size_impl;
+            observation.state = ObservationState::Done;
+            let target_depth = calculate_depth_for_node(target);
+            if target_depth < *shallowest_target_depth {
+                *shallowest_target_depth = target_depth;
+            }
+        }
+        let _ = self
+            .callback
+            .Call__(entries, self, ExceptionHandling::Report);
```

Comment:
Should this only happen when the size changes? I'm a bit unclear on the mechanism here because is_active() checks to see if the size has changed, but it seems this code both updates the size and then calls the resize observer callbacks. I don't see how the code avoids that if the size changes.

It also seems like `is_active()` is calculating the size but not caching it anywhere...meaning it updates here as well. It's not a cheap calculating right now, unfortunately, so that seems a little wasteful. Perhaps I misunderstand this code though...?

Reply:
> Should this only happen when the size changes?

This only happens for active observations, so that means size has changed plus observation depth is right. The code here does not update the size.

> It's not a cheap calculating right now, unfortunately, so that seems a little wasteful.

I have added a caching mechanism, see the last commit.